File: Csharp/DomainObjects/Domain/rule/ReserveChangeRule.cs
using System;
using System.Collections.Generic;
using System.Xml.Schema;

namespace DomainObjects.Domain.rule
{
    /// <summary>
    /// ステータス変更ルールの表条件
    /// </summary>
    public class ReserveChangeRule
    {
        Dictionary<ReservationStatus, HashSet<ReservationStatus>> map = new Dictionary<ReservationStatus,HashSet<ReservationStatus>>();

        public ReserveChangeRule()
        {
            define(ReservationStatus.仮予約, Enum.GetValues() .仮予約, ReservationStatus.キャンセル済み,]);
            map.Add(ReservationStatus.仮予約, new HashSet<ReservationStatus>() {ReservationStatus.予約済み});
            map.Add(ReservationStatus.予約済み, new HashSet<ReservationStatus>() {ReservationStatus.キャンセル済み});
            map.Add(ReservationStatus.キャンセル済み, new HashSet<ReservationStatus>() {});
        }
        
        

        public bool CanChange(ReservationStatus from, ReservationStatus to)
        {
            var allowedStatus = map[from];
            return allowedStatus.Contains(to);
        }
    }

    public enum ReservationStatus
    {
        仮予約,
        予約済み,
        キャンセル済み,
    }
}

File: Python/domain/reserve_change_rule.py
import enum
from dataclasses import dataclass
from typing import Dict, Set


class ReservationStatus(enum.Enum):
    仮予約 = enum.auto()
    予約済み = enum.auto()
    キャンセル済み = enum.auto()

@dataclass
class ReserveChangeRule:
    map: Dict[ReservationStatus, Set[ReservationStatus]]

    def __init__(self):
        self.map = {
            ReservationStatus.仮予約: {ReservationStatus.予約済み, },
            ReservationStatus.予約済み: {ReservationStatus.キャンセル済み, },
            ReservationStatus.キャンセル済み: {},
        }

    def CanChange(self, _from, _to) -> bool:
        return _to in self.map[_from]
